envoy: Support LB capability for existing k8s Service

sayboras opened the pull request:
### Description

Please refer to individual commit for more details

### Checklist

- [x] Add mutation functions based on annotation for customized and advanced use cases
- [x] Unit test
- [x] Different service types
    - [x] Node port
    - [x] LB
    - [x] ClusterIP
- [x] Support gRPC load-balancing
- [x] Integration with clustermesh
- [x] Getting started guide and docs, this can be done after first round or review or just subsequent PR. _Done as part of https://github.com/cilium/cilium/pull/23100_

### Testing

Please find below one of test cases with ClusterIP

<details>
<summary>Deployment manifest used for testing</summary>

```
$ cat temp.yaml
---
apiVersion: v1
kind: Service
metadata:
  name: rebel-base
spec:
  type: ClusterIP
  ports:
    - port: 80
  selector:
    name: rebel-base
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: rebel-base
spec:
  selector:
    matchLabels:
      name: rebel-base
  replicas: 2
  template:
    metadata:
      labels:
        name: rebel-base
    spec:
      containers:
        - name: rebel-base
          image: docker.io/nginx:1.15.8
          volumeMounts:
            - name: html
              mountPath: /usr/share/nginx/html/
          livenessProbe:
            httpGet:
              path: /
              port: 80
            periodSeconds: 1
          readinessProbe:
            httpGet:
              path: /
              port: 80
      volumes:
        - name: html
          configMap:
            name: rebel-base-response
            items:
              - key: message
                path: index.html
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: rebel-base-response
data:
  message: "{\"Galaxy\": \"Alderaan\", \"Cluster\": \"Cluster-1\"}\n"
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: x-wing
spec:
  selector:
    matchLabels:
      name: x-wing
  replicas: 1
  template:
    metadata:
      labels:
        name: x-wing
    spec:
      containers:
        - name: x-wing-container
          image: docker.io/cilium/json-mock:1.2
          livenessProbe:
            exec:
              command:
                - curl
                - -sS
                - -o
                - /dev/null
                - localhost
          readinessProbe:
            exec:
              command:
                - curl
                - -sS
                - -o
                - /dev/null
                - localhost

```

</details>


<details>
<summary>No L7 LB annotation</summary>

```
# Before apply L7 LB annotation, request is going directly to backend
$ kex deployment/x-wing -- curl -v http://rebel-base:80/
*   Trying 10.110.111.135...
* TCP_NODELAY set
* Connected to rebel-base (10.110.111.135) port 80 (#0)
> GET / HTTP/1.1
> Host: rebel-base
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.15.8
< Date: Mon, 12 Sep 2022 14:37:44 GMT
< Content-Type: text/html
< Content-Length: 47
< Last-Modified: Mon, 12 Sep 2022 14:37:40 GMT
< Connection: keep-alive
< ETag: "631f4434-2f"
< Accept-Ranges: bytes
< 
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}
* Curl_http_done: called premature == 0
* Connection #0 to host rebel-base left intact

```

</details>


<details>
<summary>With L7 LB annotation</summary>

```
# After apply L7 LB annotation, request is redirected to envoy and then follow LB policy configured
$ kg service rebel-base -o json | jq .metadata.annotations
{
  "io.cilium/global-service": "true",
  "io.cilium.service/lb-l7": "enabled",
...
}

$ kex deployment/x-wing -- curl -v http://rebel-base:80/
*   Trying 10.103.190.27...
* TCP_NODELAY set
* Connected to rebel-base (10.103.190.27) port 80 (#0)
> GET / HTTP/1.1
> Host: rebel-base
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< server: envoy
< date: Wed, 21 Dec 2022 05:03:24 GMT
< content-type: text/html
< content-length: 47
< last-modified: Wed, 21 Dec 2022 04:59:15 GMT
< etag: "63a292a3-2f"
< accept-ranges: bytes
< x-envoy-upstream-service-time: 0
< 
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}

```

</details>


<details>
<summary>global ports</summary>

```
$ kg cec
NAME                              AGE
cilium-envoy-lb-rebel-base-8080   36s

# No annotation
$ kg services rebel-base-8080 -o json | jq .metadata.annotations
{
  "io.cilium/global-service": "true",
}

$ kex deployment/x-wing -- curl -v http://rebel-base-8080:8080
* Rebuilt URL to: http://rebel-base-8080:8080/
*   Trying 10.107.97.254...
* TCP_NODELAY set
* Connected to rebel-base-8080 (10.107.97.254) port 8080 (#0)
> GET / HTTP/1.1
> Host: rebel-base-8080:8080
> User-Agent: curl/7.52.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< server: envoy
< date: Sun, 18 Sep 2022 12:15:11 GMT
< content-type: text/html
< content-length: 47
< last-modified: Sun, 18 Sep 2022 12:13:37 GMT
< etag: "63270b71-2f"
< accept-ranges: bytes
< x-envoy-upstream-service-time: 0
< 
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}
* Curl_http_done: called premature == 0
* Connection #0 to host rebel-base-8080 left intact

```

</details>

sayboras committed:
envoy: Add skeleton for service LB via envoy

This commit contains just a simple skeleton for L7 load balancing
capability via envoy proxy. There is no processing logic at all
right now.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
cli: Add configuration flag for service LB

Just to add a CLI flag along with how to configure this flag from helm

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
envoy: Add L7 service loadbalancing capability

This commit is to add a first cut to support l7 service lb capability
via envoy. The base configuration is generated from service spec, more
advance features via annotation will be added in subsequent commits.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
envoy: Add upgrade configuration for websocket

This commit is to add simple mutation function for grpc service.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

sayboras committed:
envoy: Add annotation support for lb mode

This commit is to add simple mutation function to set LB mode in envoy
cluster. By default, round-robin mode is used.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1332,6 +1332,10 @@ monitor:
 
+  # -- Enable service load balancing via envoy proxy
+  # This will automatically set enable-envoy-config as well.
+  # envoy: false
```
Helm docs unfortunately seems unable to parse those commented out values. Could you set it to a sensible default such that it is documented?

sayboras replied:
good point :100:, I will also keep an eye on below PR

https://github.com/cilium/cilium/pull/21195

sayboras replied:
update the helm values with default value :100:

gandro reviewed (commented):
Reviewed Helm only

squeed commented on operator/pkg/service/doc.go:
```diff
@@ -0,0 +1,14 @@
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const Subsys = "envoy-lb-controller"
```
This is a bit surprising; it's not clear why `operator/pkg/service` has `envoy` in the subsystem name.

sayboras replied:
my bad, ciliumenvoyconfig seems to be a better name here, it's also aligning with other packages in the same dir (e.g. ingress, ciliumenpointslice)

squeed commented on operator/pkg/service/envoy_listener.go:
```diff
@@ -0,0 +1,150 @@
+		queue:      workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		maxRetries: 10,
+	}
+	manager.serviceStore, manager.serviceInformer = informer.NewInformer(
```
I'm not super familiar with the operator structure yet, but can this use a shared informer?

christarazi replied:
The current operator structure doesn't have the most modern way of handling K8s watches. Right now, we have central watchers (single Informer) which then fan out to "subscribers". You can see this in `operator/watchers/k8s_service_sync.go`. If we make this `EnvoyLBController` a shared informer, then we can slowly improve the operator code around K8s watch handling. If that isn't possible, then I'd suggest following the "subscriber" pattern as shown in the file I mentioned earlier. 

We certainly don't want to have two duplicated K8s service watchers running for no reason.

sayboras replied:
Thanks, I have refactored the code to use subscriber pattern, one catch is that I didn't use service cache, as each annotation might need to map to one attribute, but we don't know how many annotations we will support right now. I would anticipate like a bunch (10+)

christarazi replied:
The current code is still creating a new informer so it will duplicate the K8s watching that also happens in `operator/watchers/k8s_service_cache.go`. To prevent this, the informer from `operator/watchers/k8s_service_cache.go` should be used instead and the "subscriber" becomes the "operator/pkg/service/envoy_listener.go".

sayboras replied:
does serviceInitOnce help to make sure we initialise informer once only ?

christarazi replied:
@sayboras Yes

squeed commented on operator/pkg/service/envoy_listener.go:
```diff
@@ -0,0 +1,150 @@
+	defer elb.queue.ShutDown()
+	go elb.serviceInformer.Run(wait.NeverStop)
+	if !cache.WaitForCacheSync(wait.NeverStop, elb.serviceInformer.HasSynced) {
+		return
```
Should probably bubble up some kind of error here.

squeed commented on operator/pkg/service/envoy_listener.go:
```diff
@@ -0,0 +1,150 @@
+	err := elb.handleEvent(event)
+	if err == nil {
+		elb.queue.Forget(event)
+	} else if elb.queue.NumRequeues(event) < elb.maxRetries {
```
should log `err` here.

squeed commented on operator/pkg/service/envoy_listener.go:
```diff
@@ -0,0 +1,150 @@
+		utils.ListerWatcherFromTyped[*slim_corev1.ServiceList](k8s.WatcherClient().CoreV1().Services(corev1.NamespaceAll)),
+		&slim_corev1.Service{},
+		0,
+		cache.ResourceEventHandlerFuncs{
```
So, the modern way of writing these sorts of event queues is to not enqueue any kind of event. Rather, just enqueue the NamespacedName of the service (or whatever) that has changed. Then, the event loop retrieves the service (from the informer) and makes any changes.

This does rely on the ultimate logic being level driven, since you don't get a diff.

This has the advantage of coalescing multiple updates to the same object if its sitting in the queue.

sayboras replied:
good point :100: , I remember that controller-runtime is following the above approach by default :100:

squeed commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1332,6 +1332,10 @@ monitor:
   # endpoints filtering
   # serviceTopology: false
 
+  # -- Enable service load balancing via envoy proxy
```
so, like, this enables the _option_ but not for all services by default, right? Can we clarify this?

When would we ever want this disabled?

christarazi commented on operator/pkg/service/annotations.go:
```diff
@@ -36,3 +36,39 @@ func GetLBProtocolAnnotation(obj metav1.Object) string {
+
+//
+// HTTPConnectionManagerMutator functions
+//
```
Do these comments need to be checked in?

christarazi commented on operator/pkg/service/annotations.go:
```diff
@@ -37,10 +39,22 @@ func GetLBProtocolModelAnnotation(obj metav1.Object) string {
 
-//
 // ClusterMutator functions
 //
```
Unnecessary diff between these two commits

rolinh commented on pkg/k8s/apis/cilium.io/v2/cec_types.go:
```diff
@@ -88,7 +88,7 @@ type Service struct {
 	// In CiliumClusterwideEnvoyConfig namespace defaults to "default".
 	// +kubebuilder:validation:Optional
-	Namespace string `json:"namespace"`
+	Namespace string `json:"namespace,omitempty"`
```
Technically, this is a breaking change. What's the reason to add this option?

sayboras replied:
there is no particular reason, I think it's making sense to have omitempty for optional field. 

I am not 100% if it's actually breaking change, but let me just drop this commit.

rolinh replied:
> I am not 100% if it's actually breaking change

I think it is. If some events consumer expect the field to always be present and act on it, they will break once it's not present anymore. But then again, I'm not exactly sure which consumer that would be.

chancez commented on operator/pkg/service/envoy_config.go:
```diff
@@ -43,14 +56,271 @@ func newEnvoyConfigManager(maxRetries int) (*envoyConfigManager, error) {
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", ciliumEnvoyLBPrefix, svc.GetName()),
+			Namespace: svc.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{
```
Should this set `Controller: true`  and/or `BlockOwnerDeletion`?  It's been a while since I've dealt with operators/controllers, but I think that those normally should be set?

sayboras replied:
I don't think controller: true is should be used here, as normally it's for another intermediate controller/resource such as ReplicaSet. 

I think we can skip BlockOwnerDeletion to have faster cleanup for parent resource (e.g. Service)

chancez commented on operator/pkg/service/annotations.go:
```diff
@@ -36,3 +36,39 @@ func GetLBProtocolAnnotation(obj metav1.Object) string {
+// grpcHttpConnectionManagerMutator returns a function that mutates the upgradeConfigs for grpc protocol
+func grpcHttpConnectionManagerMutator(obj metav1.Object) httpConnectionManagerMutator {
+	return func(manager *http_connection_manager_v3.HttpConnectionManager) *http_connection_manager_v3.HttpConnectionManager {
+		if GetLBProtocolAnnotation(obj) == "grpc" {
```
Is it possible to automatically configure the `UpgradeConfig` based on the incoming request instead of requiring it to be explicitly enabled?

sayboras replied:
oh right, I remembered that you mentioned this point before, totally forgot about it. Let me check and get back to you later.

tommyp1ckles commented on operator/pkg/service/envoy_config.go:
```diff
@@ -0,0 +1,56 @@
+	}
+	envoyConfig, ok := objFromCache.(*v2.CiliumEnvoyConfig)
+	if !ok {
+		return nil, exists, fmt.Errorf("got invalid object from cache")
```
I like to use the %T format string for these i.e.

```go
fmt.Errorf("got invalid object from cache: %T", objFromCache)
```

sayboras replied:
good point thanks :100:

tommyp1ckles commented on operator/pkg/service/envoy_config.go:
```diff
@@ -43,14 +56,271 @@ func newEnvoyConfigManager(maxRetries int) (*envoyConfigManager, error) {
+}
+
+func getClusterResources(svc *slim_corev1.Service) ([]ciliumv2.XDSResource, error) {
+	cluster := &envoy_config_cluster_v3.Cluster{
```
Question: Is there eventually going to be a way to configure lower level Envoy configs via the Service resource?

sayboras replied:
I think this PR is kind of a first step to achieve the above i.e. configuring envoy resource from k8s service annotation. 

I just add 2 simple mutation methods based on service annotation, but we can easily add more.

tommyp1ckles commented on operator/pkg/service/envoy_config.go:
```diff
@@ -43,14 +56,271 @@ func newEnvoyConfigManager(maxRetries int) (*envoyConfigManager, error) {
+				UpstreamProtocolOptions: &envoy_config_upstream.HttpProtocolOptions_UseDownstreamProtocolConfig{
+					UseDownstreamProtocolConfig: &envoy_config_upstream.HttpProtocolOptions_UseDownstreamHttpConfig{
+						// Empty HTTP/2 options has no effect, so this should not be needed
+						Http2ProtocolOptions: &envoy_config_core_v3.Http2ProtocolOptions{},
```
This field is marked as deprecated in the spec, will the API not accept it without this?

sayboras replied:
good point, let me update to newer API

sayboras replied:
actually, the go struct we are using didn't show deprecated warning, can you share me more details/reference?

tommyp1ckles replied:
I believe it was in the proto spec files

tommyp1ckles replied:
I don't see it either, I might've referenced the wrong line. Let me figure out what this was in reference too.

tommyp1ckles replied:
Not sure where I saw that 🤷‍♂️ , all your protobuf structs look good to me.

sayboras committed:
envoy: Support list of global ports for redirect

This commit is to support envoy redirect for all services having port
number in list, so that user didn't need to annotate one by one. Just a
note that for more customized and advanced use case, service annotation
will be still required.

One caveat is that the current CEC operates in service level i.e. there
is no granular control to only redirect one particular service port.

Signed-off-by: Tam Mach <tam.mach@cilium.io>

squeed commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1338,6 +1338,18 @@ monitor:
+    # Please refer to docs for supported annotations for configuring the envoy proxy.
+    # This will automatically set enable-envoy-config as well.
+    enabled: false
+    # -- List of ports from service to be automatically redirected to envoy proxy.
```
Does this apply to all services, or only those with the annotation?

sayboras replied:
this will be applied for all services, good spot :eyes:. I will update comment to make it clearer.

squeed replied:
Hmm, what's the use-case for that? It it seems like a bit of a footgun.

squeed replied:
(we discussed out-of-band, just need to make this clear that this applies to **all services** in the cluster)

gandro commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -1335,6 +1335,18 @@ monitor:
+    enabled: false
+    # -- List of ports from service to be automatically redirected to envoy proxy.
+    # Fine-grained control can be achieved by using the service annotation.
+    ports: [ ]
```
nit (to be consistent with the rest of the file):
```suggestion
    ports: []
```

gandro reviewed (approved):
Helm related changes look good to me, only a minor nit

sayboras committed:
envoy: Support default LB algorithm flag

The default load balancer algorithm can still be overridden by service
annotation (e.g. io.cilium.service/lb-algorithm).

Signed-off-by: Tam Mach <tam.mach@cilium.io>

chancez commented on operator/pkg/ciliumenvoyconfig/envoy_config.go:
```diff
@@ -0,0 +1,340 @@
+		OutlierDetection: &envoy_config_cluster_v3.OutlierDetection{
+			SplitExternalLocalOriginErrors: true,
+			// The number of consecutive locally originated failures before ejection occurs.
+			ConsecutiveLocalOriginFailure: &wrapperspb.UInt32Value{Value: 2},
```
The default is 5, why do we pick 2?

sayboras replied:
good point :100:, it's better to just stick with the default right now :100:

chancez commented on operator/pkg/ciliumenvoyconfig/envoy_config.go:
```diff
@@ -0,0 +1,340 @@
+				},
+			}),
+		},
+		OutlierDetection: &envoy_config_cluster_v3.OutlierDetection{
```
The default is that this is false, why does this need to be configured? Perhaps some comments above could describe the reason we configure OutlierDetection

sayboras replied:
same as above, I probably tested with different config, let's stick with the default and then support customization based on users' feedback.

chancez commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+		if IsLBProtocolAnnotationEnabled(service) || hasAnyPort(service, m.ports) {
+			return m.createEnvoyConfig(ctx, service)
+		}
+		return m.deleteEnvoyConfig(ctx, service)
```
This is a bit confusing IMO, I think we should probably have a condition where we call delete, instead of it being the fallback. Is this intending to delete the config when the service is deleted? Perhaps we should check the deletionTimestamp before deleting.

sayboras replied:
this is for the case user add or remove or update the annotation.

chancez reviewed (commented):
Overall seems solid, most comments are nit-picks. One question: would this approach allow a future enhancement where we could run a dedicated pool load balancer nodes?

christarazi commented on operator/pkg/ciliumenvoyconfig/annotations.go:
```diff
@@ -0,0 +1,27 @@
+package ciliumenvoyconfig
```
Missing copyright header

christarazi commented on operator/watchers/cec.go:
```diff
@@ -0,0 +1,38 @@
+
+	swgSvcs := lock.NewStoppableWaitGroup()
+	swgEps := lock.NewStoppableWaitGroup()
+	InitServiceWatcher(cfg, clientset, swgSvcs, swgEps, optsModifier)
```
Given that the `serviceInitOnce` prevents multiple informer instantiations, the `optsModifier` here won't have an effect because `StartCECController` will probably be called after the service informer is first initialized. Is that a problem?

sayboras replied:
this is no longer applicable as per that latest changes on cached resource (e.g. legacy.resource.Services)

christarazi commented on operator/pkg/ciliumenvoyconfig/envoy_config.go:
```diff
@@ -129,10 +130,15 @@ func (m *manager) getResources(svc *slim_corev1.Service) ([]ciliumv2.XDSResource
 }
 
 func (m *manager) getClusterResources(svc *slim_corev1.Service) ([]ciliumv2.XDSResource, error) {
+	lbPolicy, ok := envoy_config_cluster_v3.Cluster_LbPolicy_value[strings.ToUpper(m.algorithm)]
+	if !ok {
+		lbPolicy = int32(envoy_config_cluster_v3.Cluster_ROUND_ROBIN)
+	}
```
Should the user be warned if they put an unacceptable algorithm and we fallback to round robin?

squeed commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1382,6 +1382,24 @@ monitor:
   # serviceTopology: false
 
+  # -- LoadBalancer with envoy proxy
+  envoy:
```
Rather than calling this section `envoy`, what about something like `layer-7`?

```yaml
loadBalancer:
    layer7:
        backend: envoy | disabled | my-cool-new-proxy
```

sayboras replied:
good point, I just check with other places, we are using `l7` term. For another level of backend, I think it will be a long road as we don't have a good integration right now to inject actual endpoints except envoy.

Updated: actually this makes other configuration param easier :100:

squeed commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -209,6 +209,15 @@ data:
 
+{{- if hasKey .Values "loadBalancer" }}
+{{- if .Values.loadBalancer.envoy.enabled }}
+  enable-loadbalancer-envoy: "true"
```
likewise, what if the cilium-config keys were called something like
```
loadbalancer-l7: envoy | disabled | traefik (or whatever)
loadbalancer-l7-ports: ...
loadbalancer-l7-algorithm: ...
```

squeed commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	if !IsLBProtocolAnnotationEnabled(service) && !hasAnyPort(service, m.ports) {
+		return nil
```
What if the service had l7lb disabled while the controller was being upgraded? If so, we'll leak the config here. You can't shortcut this case :-)

sayboras replied:
can you elaborate a little bit more here ? This is for service added event, also the annotation will have higher priority i.e. if the service annotation is disabled, we will not do anything even the port requirement is satisfied.

squeed commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+
+	scopedLog := log.WithField(logfields.ServiceKey, getName(svc))
+	if exists {
+		if desired.DeepEqual(existingEnvoyConfig) {
```
Does it make sense to exclude Status from this DeepEqual?

sayboras replied:
This is good point, however we don't have any status as such in CEC now :(, I agreed that we should have some status in CEC, however, we didn't really think on how to maintain status (i.e. per node vs all node)

squeed commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+		// Update existing CEC
+		newEnvoyConfig := existingEnvoyConfig.DeepCopy()
+		newEnvoyConfig.Spec = desired.Spec
+		_, err = k8s.CiliumClient().CiliumV2().CiliumEnvoyConfigs(svc.Namespace).Update(context.Background(), newEnvoyConfig, metav1.UpdateOptions{})
```
Is it worth doing a Patch here, or is that an unnecessary optimization?

chancez replied:
I wondered about this, if something else is updating the CEC then this might break that.

squeed commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+		return nil
+	}
+
+	_, err = k8s.CiliumClient().CiliumV2().CiliumEnvoyConfigs(svc.Namespace).Create(ctx, desired, metav1.CreateOptions{})
```
Might as well set a FieldManager here (and for Update too). Can't hurt.

squeed commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+	return nil
+}
+
+func (m *manager) deleteEnvoyConfig(ctx context.Context, svc *slim_corev1.Service) error {
```
Can we shortcut this case if the CEC isn't in the store?

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/annotations.go:
```diff
@@ -0,0 +1,27 @@
+
+// IsLBProtocolAnnotationEnabled returns true if the load balancer protocol is enabled
+func IsLBProtocolAnnotationEnabled(svc *slim_corev1.Service) bool {
+	return svc.GetAnnotations()[lbProtocolAnnotation] != ""
```
do we not need to check if its not "false"

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/envoy_config.go:
```diff
@@ -258,7 +258,9 @@ func getConnectionManager(svc *slim_corev1.Service) (ciliumv2.XDSResource, error
 
-	var mutatorFuncs = []httpConnectionManagerMutator{}
+	var mutatorFuncs = []httpConnectionManagerMutator{
+		grpcHttpConnectionManagerMutator(svc),
```
I'm a bit confused by this, this returns a function which is-a mutator, why do we need to pass the svc which is then unused?

sayboras replied:
it's related to above comment from Chance. We are not configuring auto upgrade for websocket, hence we don't need to check if it's grpc anymore.

Agreed that input for this function is not used, this is just to keep the same mutator function signature, so that we can add more features later.

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/annotations_test.go:
```diff
@@ -0,0 +1,30 @@
+				},
+			},
+		})(input)
+		require.NotNil(t, res)
```
Is there a reason to use require for these, I think /testify/assert is much more common in the code base (23 vs 72 hits)

sayboras replied:
AFAIK, assert will not stop the test immediately, hence let's say this assert if false (i.e. res is nil), the next statement will cause panic.

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/envoy_config.go:
```diff
@@ -0,0 +1,56 @@
+	return manager, nil
+}
+
+// getByKey is a wrapper of Store.GetByKey but with concrete Endpoint object
```
think you want to change this Endpoint -> CiliumEnvoyConfig

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,86 @@
+	manager := &manager{
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		serviceStore: indexer,
+		maxRetries:   10,
```
should this be configurable (jw, grepping through cilium it seems like most similar cases are configured it some opts struct)

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,86 @@
+		}
+		err := m.processEvent(ev)
+		if err != nil {
+			if m.queue.NumRequeues(ev) < m.maxRetries {
```
this snippet seems to get repeated a lot in cilium, wonder if we should factor this out someday.

sayboras replied:
yup, I just came to know the new generic pattern added recently. I think after sometime, we can adopt it widely.

tommyp1ckles commented on operator/pkg/ciliumenvoyconfig/envoy_config.go:
```diff
@@ -58,3 +71,260 @@ func (em *envoyConfigManager) getByKey(key string) (*v2.CiliumEnvoyConfig, bool,
+			"envoy.extensions.upstreams.http.v3.HttpProtocolOptions": toAny(&envoy_config_upstream.HttpProtocolOptions{
+				UpstreamProtocolOptions: &envoy_config_upstream.HttpProtocolOptions_UseDownstreamProtocolConfig{
+					UseDownstreamProtocolConfig: &envoy_config_upstream.HttpProtocolOptions_UseDownstreamHttpConfig{
+						// Empty HTTP/2 options has no effect, so this should not be needed
```
Why do we add it if its not needed?

chancez commented on operator/pkg/ciliumenvoyconfig/manager.go:
```diff
@@ -0,0 +1,231 @@
+	}
+	manager.envoyConfigManager = envoyConfigManager
+
+	go manager.run(ctx)
```
I think in general, Cilium is trying to move away from running Go routines in the constructor, so this could be moved into a `manager.Run` which the caller is expected to start.
